### PR TITLE
Fix ROVIOLI multi camera localization mode

### PR DESCRIPTION
**Pull request by multiagent-mapping-sheep**
**Problem:**

If ROVIOLI is using more than one camera in ncamera.yaml, the underlying ROVIO is only using the first camera for the VIO, however localization and map building are working with all available cameras. This was fine for the original localization implementation, that used 6DoF constraints to incorporate any localization results in ROVIO, however, with the new localization constraints, based on 2D-3D matches this resulted in a bug. The localizer was providing ROVIO with constraints for cameras that it was not aware off, which, depending on the version, lead to either crashes, or no localizations. 

**Changes:**

 - [x] This patch makes ROVIO discard all localization results that originate from cameras it's not tracking, unless the 6DoF constraint localization mode is used. 
 - [x] In order to leverage the fact that in non-overlapping multi-camera setups other cameras might localize in places the primary tracking camera does not, we introduce the following feature: If there is a localization, for any camera but the camera used by ROVIO, the localization result can be applied to the filter as a 6DoF constraint. Since the main reason for switching from 6DoF to structure constraints was the improved covariance, this feature is optional and by default disabled.
Flag: `--rovioli_use_6dof_localization_for_inactive_cameras`
 - [x] A flag to apply a time offset between IMU and camera. This is a small feature that has been requested several times and can help to deal with sensors that lack proper IMU to camera synchronization, but have a more or less stable offset.
 - [x] Minor changes in the ROVIO log output.

The following multi-camera configurations and localization map configurations have been successfully tested on the EUROC dataset:

| Query: EUROC MH 2 Localization map: EUROC MH 1 	| Stereo ROVIOLI  (LOC-mode, use 2nd camera as 6DoF constraints) 	| Stereo ROVIOLI (LOC-mode, discard 2nd camera results) 	| Monocular ROVIOLI (LOC-mode) 	|
|------------------------------------------------	|----------------------------------------------------------------	|-------------------------------------------------------	|------------------------------	|
| Localization map from stereo camera 	| X 	| X 	| X 	|
| Localization map from monocular camera 	| X 	| X 	| X 	|

Credits: @mfehr, @schneith

### CODE REVIEW

**Comment by mfehr** [comment]
test this please (jenkins config was wrong, did not use the dependencies.rosinstall yet)

File: applications/rovioli/include/rovioli/rovio-localization-handler.h [hunk 1]
```diff
@@ -54,7 +56,8 @@ class RovioLocalizationHandler {
   bool processAsUpdate(
       const vio::LocalizationResult::ConstPtr& localization_result);
 
-  bool getLocalizationReprojectionErrors(
+  // Returns the ratio of successfully reprojected matches
```
Comment (schneith) [inline applications/rovioli/include/rovioli/rovio-localization-handler.h]:
```suggestion
  // Returns the ratio of successfully reprojected matches.
```

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 2]
```diff
@@ -246,13 +261,65 @@ bool RovioLocalizationHandler::processAsUpdate(
     measurement_accepted = rovio_interface_->processGroundTruthUpdate(
         JrJV, qJV, rovio_timestamp_sec);
   } else {
+    // Check if there are any matches to be processed in the camera frames that
+    // are used by ROVIO for estimation (inactive).
+    const size_t num_cameras =
+        localization_result->G_landmarks_per_camera.size();
+    size_t num_valid_matches = 0u;
+    for (size_t cam_idx = 0u; cam_idx < num_cameras; ++cam_idx) {
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]:
```suggestion
    for (size_t maplab_cam_idx = 0u; maplab_cam_idx < num_cameras; ++maplab_cam_idx) {
```

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 3]
```diff
@@ -246,13 +261,65 @@ bool RovioLocalizationHandler::processAsUpdate(
...
+      const size_t* rovio_cam_idx =
+          maplab_to_rovio_cam_indices_mapping_.getRight(cam_idx);
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]:
```suggestion
          maplab_to_rovio_cam_indices_mapping_.getRight(maplab_cam_idx);
```

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 4]
```diff
@@ -246,13 +261,65 @@ bool RovioLocalizationHandler::processAsUpdate(
...
+      if (rovio_cam_idx == nullptr) {
+        continue;
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]: // Camera inactive in ROVIO.

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 5]
```diff
@@ -246,13 +261,65 @@ bool RovioLocalizationHandler::processAsUpdate(
...
+      }
+      num_valid_matches +=
+          localization_result->G_landmarks_per_camera[cam_idx].cols();
+    }
+    if (num_valid_matches == 0u) {
+      // There are no valid localization matches for the cameras used by ROVIO.
+      // Use the localization result of the inactive cameras but integrate them
+      // using the 6dof localization mode, since we cannot currently pass 2D-3D
+      // correspondences to ROVIO for a camera it's not using.
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]: Maybe we can just initialise the camera but not pass images but still 2d-3d constraints? I'll take a look if that is possible later.

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 6]
```diff
@@ -246,13 +261,65 @@ bool RovioLocalizationHandler::processAsUpdate(
...
+            JrJV, qJV, rovio_timestamp_sec);
+
+        VLOG_IF(1, measurement_accepted)
+            << "No localization found for active camera, successfully updated "
+            << "ROVIO using 6DoF constraints based on localization from "
+            << "inactive cameras.";
+
+        LOG_IF(
+            WARNING, !measurement_accepted && rovio_interface_->isInitialized())
+            << "No localization found for active camera, failed to updated "
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]:
```suggestion
            << "No localization found for active camera, failed to update "
```

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 7]
```diff
@@ -278,31 +346,33 @@ bool RovioLocalizationHandler::processAsUpdate(
...
             << "the image plane. Will reset the localization.";
       }
-      localization_state_ = LocalizationState::kNotLocalized;
-      T_G_M_lc_buffer_.clear();
-      return initializeBaseframe(localization_result);
     }
 
-    const size_t num_cameras =
-        localization_result->G_landmarks_per_camera.size();
     for (size_t cam_idx = 0u; cam_idx < num_cameras; ++cam_idx) {
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]: maplab_cam_idx

File: applications/rovioli/src/rovio-localization-handler.cc [hunk 8]
```diff
@@ -357,10 +440,9 @@ bool RovioLocalizationHandler::getLocalizationReprojectionErrors(
   }
 
   CHECK_EQ(lc_reprojection_errors->size(), filter_reprojection_errors->size());
-
-  constexpr double kMinSuccessfulReprojectionRatio = 0.5;
-  return (static_cast<double>(lc_reprojection_errors->size()) /
-          num_matches_processed) > kMinSuccessfulReprojectionRatio;
+  CHECK_GT(num_matches_processed, 0);
+  return static_cast<double>(lc_reprojection_errors->size()) /
```
Comment (schneith) [inline applications/rovioli/src/rovio-localization-handler.cc]: Shouldn't we also check the absolute value?

File: applications/rovioli/src/rovio-localization-handler.cc (hunk 5 above)
Comment (mfehr) [reply applications/rovioli/src/rovio-localization-handler.cc]: I agree, this is clearly the better solution, but let's keep this for now, it's anyways optional.

File: applications/rovioli/src/rovio-localization-handler.cc (hunk 8 above)
Comment (mfehr) [reply applications/rovioli/src/rovio-localization-handler.cc]: We could add an additional check here as well, but wouldn't the localizer fail if the number of matches is too low and there wouldn't be a result here to process anyways?

**Comment by mfehr** [comment]
@schneith  thanks for the review

**Pull request merged**